File: app/src/main/java/pl/karol202/paintplus/options/OptionLayerOpen.java
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package pl.karol202.paintplus.options;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.media.ExifInterface;
import android.support.design.widget.Snackbar;
import pl.karol202.paintplus.AsyncManager;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.activity.ActivityPaint;
import pl.karol202.paintplus.file.UriMetadata;
import pl.karol202.paintplus.history.action.ActionLayerAdd;
import pl.karol202.paintplus.image.Image;
import pl.karol202.paintplus.image.Image.FlipDirection;
import pl.karol202.paintplus.image.layer.Layer;

public class OptionLayerOpen extends OptionOpen
{
	private static final int REQUEST_OPEN_LAYER = 3;
	
	public OptionLayerOpen(ActivityPaint activity, Image image, AsyncManager asyncManager)
	{
		super(activity, image, asyncManager);
	}
	
	@Override
	int getRequestId()
	{
		return REQUEST_OPEN_LAYER;
	}
	
	@Override
	public void onImageLoaded(Bitmap bitmap)
	{
		if(bitmap == null) getAppContext().createSnackbar(R.string.message_cannot_open_file, Snackbar.LENGTH_SHORT).show();
		else addNewLayer(bitmap);
	}
	
	private void addNewLayer(Bitmap bitmap)
	{
		final Layer layer = new Layer(0, 0, bitmap.getWidth(), bitmap.getHeight(), getLayerName(), Color.TRANSPARENT);
		layer.setBitmap(bitmap);
		if(!getImage().addLayer(layer, 0)) getAppContext().createSnackbar(R.string.too_many_layers, Snackbar.LENGTH_SHORT).show();
		else
		{
			createLayerAddHistoryAction(layer);
			askAboutExifRotation(new RotationNeedListener() {
				@Override
				public void onRotationNeed(int exifOrientation)
				{
					rotateLayer(layer, exifOrientation);
				}
			});
		}
	}
	
	private String getLayerName()
	{
		UriMetadata metadata = new UriMetadata(getContext(), getUri());
		return metadata.getDisplayName();
	}
	
	private void createLayerAddHistoryAction(Layer layer)
	{
		ActionLayerAdd action = new ActionLayerAdd(getImage());
		action.setLayerAfterAdding(layer);
		action.applyAction();
	}
	
	private void rotateLayer(Layer layer, int exifOrientation)
	{
		switch(exifOrientation)
		{
		case ExifInterface.ORIENTATION_ROTATE_90:
			layer.rotate(90);
			break;
		case ExifInterface.ORIENTATION_ROTATE_180:
			layer.rotate(180);
			break;
		case ExifInterface.ORIENTATION_ROTATE_270:
			layer.rotate(270);
			break;
		case ExifInterface.ORIENTATION_FLIP_HORIZONTAL:
			layer.flip(FlipDirection.HORIZONTALLY);
			break;
		case ExifInterface.ORIENTATION_FLIP_VERTICAL:
			layer.flip(FlipDirection.VERTICALLY);
			break;
		case ExifInterface.ORIENTATION_TRANSPOSE:
			layer.rotate(90);
			layer.flip(FlipDirection.HORIZONTALLY);
			break;
		case ExifInterface.ORIENTATION_TRANSVERSE:
			layer.rotate(-90);
			layer.flip(FlipDirection.HORIZONTALLY);
		}
	}
}

File: app/src/main/java/pl/karol202/paintplus/options/OptionLayerOpen.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.options

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import pl.karol202.paintplus.R
import pl.karol202.paintplus.history.Action
import pl.karol202.paintplus.image.*
import pl.karol202.paintplus.image.layer.Layer
import pl.karol202.paintplus.util.getDisplayName
import pl.karol202.paintplus.viewmodel.PaintViewModel

class OptionLayerOpen(private val context: Context,
                      private val imageService: ImageService,
                      private val historyService: HistoryService,
                      private val effectsService: EffectsService,
                      private val openOption: OptionOpen) : Option
{
	private val actionPreset = Action.namePreset(R.string.history_action_layer_add)

	fun execute() = openOption.execute(this::onResult)

	private fun onResult(result: OptionOpen.OpenResult) = when(result)
	{
		is OptionOpen.OpenResult.Success -> onUriSelected(result.uri, result.bitmap, result.exifOrientation)
		is OptionOpen.OpenResult.Failed -> effectsService.showMessage(R.string.message_cannot_open_file)
	}

	private fun onUriSelected(uri: Uri, bitmap: Bitmap, orientation: Int?) =
			openOption.askAboutExifRotation(orientation = orientation,
			                                onRotationApply = { onUriAndRotationSelected(uri, bitmap, it) },
			                                onNoRotation = { onUriAndRotationSelected(uri, bitmap, null) })

	private fun onUriAndRotationSelected(uri: Uri, bitmap: Bitmap, orientation: Int?)
	{
		if(!imageService.image.canAddMoreLayers) return
		historyService.commitAction { commit(uri, bitmap, orientation) }
	}

	private fun commit(uri: Uri, bitmap: Bitmap, orientation: Int?): Action.ToRevert
	{
		val oldImage = imageService.image
		val layer = Layer.create(0, 0, createName(uri), bitmap).rotatedByExif(orientation)
		imageService.editImage { withLayerAdded(layer, autoSelect = true) }
		return actionPreset.toRevert(layer.bitmap) { revert(uri, bitmap, orientation, layer, oldImage) }
	}

	private fun revert(uri: Uri, bitmap: Bitmap, orientation: Int?, newLayer: Layer, oldImage: Image): Action.ToCommit
	{
		imageService.setImage(oldImage)
		return actionPreset.toCommit(newLayer.bitmap) { commit(uri, bitmap, orientation) }
	}

	private fun createName(uri: Uri) =
			uri.getDisplayName(context) ?: context.getString(R.string.opened_layer_name)

	private fun Layer.rotatedByExif(exifOrientation: Int?) = when(exifOrientation)
	{
		ExifInterface.ORIENTATION_ROTATE_90 -> rotated(90f)
		ExifInterface.ORIENTATION_ROTATE_180 -> rotated(180f)
		ExifInterface.ORIENTATION_ROTATE_270 -> rotated(270f)
		ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> flipped(FlipDirection.HORIZONTALLY)
		ExifInterface.ORIENTATION_FLIP_VERTICAL -> flipped(FlipDirection.VERTICALLY)
		ExifInterface.ORIENTATION_TRANSPOSE -> rotated(90f).flipped(FlipDirection.HORIZONTALLY)
		ExifInterface.ORIENTATION_TRANSVERSE -> rotated(-90f).flipped(FlipDirection.HORIZONTALLY)
		else -> this
	}
}
